studio/components/interfaces/TableGridEditor/SidePanelEditor/SidePanelEditor.tsx

```diff
@@ -514,66 +501,83 @@ const SidePanelEditor = ({
       message: `Successfully imported ${rowCount} rows of data into ${selectedTable.name}`,
     })
     resolve()
-    closePanel()
+    snap.closeSidePanel()
   }
 
   const onClosePanel = () => {
     if (isEdited) {
       setIsClosingPanel(true)
     } else {
-      closePanel()
+      snap.closeSidePanel()
     }
   }
 
   return (
     <>
       {!isUndefined(selectedTable) && (
         <RowEditor
-          row={selectedRowToEdit}
+          row={snap.sidePanel?.type === 'row' ? snap.sidePanel.row : undefined}
           selectedTable={selectedTable}
-          visible={sidePanelKey === 'row'}
+          visible={snap.sidePanel?.type === 'row'}
           closePanel={onClosePanel}
           saveChanges={saveRow}
           updateEditorDirty={() => setIsEdited(true)}
         />
       )}
       {!isUndefined(selectedTable) && (
         <ColumnEditor
-          column={selectedColumnToEdit}
+          column={
+            snap.sidePanel?.type === 'column'
+              ? (snap.sidePanel.column as PostgresColumn)
+              : undefined
+          }
           selectedTable={selectedTable}
-          visible={sidePanelKey === 'column'}
+          visible={snap.sidePanel?.type === 'column'}
           closePanel={onClosePanel}
           saveChanges={saveColumn}
           updateEditorDirty={() => setIsEdited(true)}
         />
       )}
       <TableEditor
-        table={selectedTableToEdit}
-        isDuplicating={isDuplicating}
-        visible={sidePanelKey === 'table'}
+        table={
+          snap.sidePanel?.type === 'table' &&
+          (snap.sidePanel.mode === 'edit' || snap.sidePanel.mode === 'duplicate')
+            ? selectedTable
+            : undefined
+        }
+        isDuplicating={snap.sidePanel?.type === 'table' && snap.sidePanel.mode === 'duplicate'}
+        visible={snap.sidePanel?.type === 'table'}
         closePanel={onClosePanel}
         saveChanges={saveTable}
         updateEditorDirty={() => setIsEdited(true)}
       />
       <JsonEdit
-        visible={sidePanelKey === 'json'}
-        column={selectedValueForJsonEdit?.column ?? ''}
-        jsonString={selectedValueForJsonEdit?.jsonString ?? ''}
+        visible={snap.sidePanel?.type === 'json'}
+        column={(snap.sidePanel?.type === 'json' && snap.sidePanel.jsonValue.column) || ''}
+        jsonString={(snap.sidePanel?.type === 'json' && snap.sidePanel.jsonValue.jsonString) || ''}
         backButtonLabel="Cancel"
         applyButtonLabel="Save changes"
         closePanel={onClosePanel}
         onSaveJSON={onSaveJSON}
       />
       <ForeignRowSelector
-        key={`foreign-row-selector-${selectedForeignKeyToEdit?.foreignKey?.id ?? 'null'}`}
-        visible={sidePanelKey === 'foreign-row-selector'}
-        foreignKey={selectedForeignKeyToEdit?.foreignKey}
+        key={`foreign-row-selector-${
```

Comment:
do we need a key prop?

Reply:
It will ensure one side panel never accidentally edits the wrong underlying data by having each side panel be it's own unique instance